File: src/02_quick_merge_sort/quick_merge_sort.js
function merge(left, right) {
    const merged = [];
    let i = 0, j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            merged.push(left[i]);
            i++;
        } else {
            merged.push(right[j]);
            j++;
        }
    }

    while (i < left.length) {
        merged.push(left[i]);
        i++;
    }

    while (j < right.length) {
        merged.push(right[j]);
        j++;
    }

    return merged;
}

function quickMergeSort(arr, threshold = 10) {
    if (arr.length <= 1) {
        return arr;
    }

    if (arr.length <= threshold) {
        const mid = Math.floor(arr.length / 2);
        const left = quickMergeSort(arr.slice(0, mid), threshold);
        const right = quickMergeSort(arr.slice(mid), threshold);
        return merge(left, right);
    }

    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);

    return [
        ...quickMergeSort(left, threshold),
        ...middle,
        ...quickMergeSort(right, threshold)
    ];
}

if (require.main === module) {
    const data = [34, 7, 23, 32, 5, 62, 32, 23, 12];
    console.log("Original:", data);
    console.log("Sorted:  ", quickMergeSort(data));
}

File: src/02_quick_merge_sort/Quick_merge_sort.py
def quick_merge_sort(arr):
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    left_sorted = quick_merge_sort(left)
    right_sorted = quick_merge_sort(right)
    return merge(left_sorted + middle, right_sorted)
def merge(left, right):
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result
if __name__ == "__main__":
    arr = [5, 2, 9, 1, 5, 6]
    print("Original:", arr)
    print("Sorted:", quick_merge_sort(arr))
